electrode-ota-server-model-app/src/app.js

```diff
@@ -123,55 +126,94 @@ export default (options, dao, upload, download) => {
         removeDeployment(params){
-            return api.findApp(params).then(app => dao.removeDeployment(app.id, params.deployment));
+            return api.findApp(params).then(app => {
+                return dao.removeDeployment(app.id, params.deployment)
+                    .tap(() => logger.info({ appId : app.id, deployment : params.deployment }, "removed deployment"));
+            });
         },
 
         renameDeployment(params){
             return api.findApp(params).then(app => {
                 const {deployment, name} = params;
                 notFound(hasDeploymentName(app, deployment), `Deployment '${deployment}' not found ${params.app}`);
                 alreadyExists(!hasDeploymentName(app, name), name, 'deployment');
-                return dao.renameDeployment(app.id, deployment, name);
+                return dao.renameDeployment(app.id, deployment, name)
+                    .tap(() => logger.info({ appId : app.id, oldName : deployment, newName : name }, "renamed deployment"));
             });
         },
 
         promoteDeployment(params)
         {
             return api.findApp(params).then(app => {
 
-                return dao.deploymentsByApp(app.id, [params.deployment, params.to]).then(deployments => {
+                return dao.deploymentsByApp(app.id, [params.deployment, params.to]).then(async (deployments) => {
                     const f = deployments[params.deployment];
 
                     notFound(f && f.package, `Deployment "${params.deployment}" does not exist.`);
 
                     const t = notFound(deployments[params.to], `Deployment "${params.to}" does not exist.`);
 
-                    const pkg = f.package;
+                    let pkg = f.package;
 
-                    const {
-                        isDisabled = pkg.isDisabled, isMandatory = pkg.isMandatory, rollout = pkg.rollout, appVersion = pkg.appVersion, description = pkg.description
-                    } = params;
+                    if (params.label) {
+                        const packages = await dao.history(app.id, [params.deployment]);
+                        pkg = packages.find((packag) => {
+                            return packag.label === params.label;
+                        });
 
+                        notFound(pkg, `Deployment "${params.deployment}" has no package with label "${params.label}"`);
+                    }
+
+                    const existingPackage = t.package;
+                    
+                    // check to make sure that it is not already promoted
+                    if (existingPackage) {
+                        console.log("from packageHash", pkg.packageHash);
+                        console.log("to packageHash", existingPackage.packageHash);
+                        alreadyExistsMsg((existingPackage.packageHash !== pkg.packageHash), 
+                            `Deployment ${params.from}:${pkg.label} has already been promoted to ${params.to}:${existingPackage.label}.`);
```

Comment:
`params.from` should be `params.deployment`

Reply:
good catch.